File: auto-commit.py
#!/usr/bin/env python3
"""
Auto-commit message generator
Analyzes git staged changes and generates meaningful commit messages
"""

import subprocess
import sys
import re
from pathlib import Path
from typing import Tuple, Dict, List


COMMIT_TYPES = {
    "feat": "A new feature",
    "fix": "A bug fix",
    "docs": "Documentation only changes",
    "style": "Changes that don't affect code meaning",
    "refactor": "Code change that neither fixes a bug nor adds a feature",
    "perf": "Code change that improves performance",
    "test": "Adding or updating tests",
    "chore": "Changes to build process or dependencies",
    "ci": "CI configuration changes",
}


def get_staged_files() -> List[str]:
    """Get list of staged files"""
    try:
        output = subprocess.check_output(
            ["git", "diff", "--cached", "--name-only"],
            text=True,
            stderr=subprocess.PIPE
        )
        return [f.strip() for f in output.split("\n") if f.strip()]
    except subprocess.CalledProcessError:
        print("Error: Not a git repository or no staged changes")
        sys.exit(1)


def get_detailed_diff() -> str:
    """Get detailed diff of staged changes"""
    try:
        output = subprocess.check_output(
            ["git", "diff", "--cached"],
            text=True,
            stderr=subprocess.PIPE
        )
        return output
    except subprocess.CalledProcessError:
        return ""


def analyze_changes(files: List[str], diff: str) -> Dict:
    """Analyze changes and categorize them"""
    stats = {
        "added_lines": len(re.findall(r"^\+", diff, re.MULTILINE)) - len(files),
        "removed_lines": len(re.findall(r"^-", diff, re.MULTILINE)) - len(files),
        "test_files": sum(1 for f in files if ".test." in f or ".spec." in f),
        "doc_files": sum(1 for f in files if f.endswith((".md", ".txt", ".rst"))),
    }

    types = set()
    for file in files:
        if "package.json" in file or "requirements.txt" in file:
            types.add("chore")
        if file.endswith((".js", ".ts", ".jsx", ".tsx", ".py", ".java")):
            types.add("feat")

    return {
        "files": files,
        "file_count": len(files),
        "stats": stats,
        "types": list(types),
    }


def generate_commit_message(analysis: Dict) -> str:
    """Generate commit message from analysis"""
    file_count = analysis["file_count"]
    stats = analysis["stats"]
    files = analysis["files"]
    types = analysis["types"]

    # Determine commit type
    commit_type = "chore"
    if stats["test_files"] > 0:
        commit_type = "test"
    elif stats["doc_files"] == file_count:
        commit_type = "docs"
    elif "feat" in types:
        commit_type = "feat"

    # Generate subject
    is_small_change = file_count <= 3

    if stats["doc_files"] == file_count:
        subject = f"{commit_type}: update documentation"
    elif stats["test_files"] > 0:
        subject = f"{commit_type}: add/update tests"
    elif stats["added_lines"] > stats["removed_lines"] * 2:
        subject = f"{commit_type}: add new functionality"
    elif stats["removed_lines"] > stats["added_lines"]:
        subject = f"{commit_type}: remove unused code"
    elif is_small_change and file_count == 1:
        subject = f"{commit_type}: update {Path(files[0]).name}"
    else:
        subject = f"{commit_type}: refactor code structure"

    # Generate body
    body = f"\nFiles changed: {file_count}\n"
    body += f"Lines added: {stats['added_lines']}\n"
    body += f"Lines removed: {stats['removed_lines']}\n"

    if file_count <= 5:
        body += "\nModified files:\n"
        for file in files:
            body += f"- {file}\n"

    return subject + body


def commit_with_message(message: str) -> None:
    """Commit with the generated message"""
    try:
        subprocess.run(
            ["git", "commit", "-m", message],
            check=True,
        )
        print("\n✓ Commit created successfully!")
    except subprocess.CalledProcessError as e:
        print(f"Error creating commit: {e}")
        sys.exit(1)


def main():
    print("📝 Analyzing staged changes...\n")

    files = get_staged_files()
    if not files:
        print("Error: No staged changes found. Stage your changes first with: git add .")
        sys.exit(1)

    diff = get_detailed_diff()
    analysis = analyze_changes(files, diff)
    message = generate_commit_message(analysis)

    print("═" * 43)
    print("✨ Generated Commit Message:")
    print("═" * 43 + "\n")
    print(message)
    print("\n" + "═" * 43)

    # Ask user if they want to commit
    try:
        answer = input("\n✅ Do you want to commit with this message? (y/n): ").strip().lower()
        if answer == "y":
            commit_with_message(message)
        else:
            print("\nCommit cancelled.")
    except KeyboardInterrupt:
        print("\n\nCommit cancelled.")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: auto-commit.js
#!/usr/bin/env node

/**
 * Auto-commit message generator
 * Analyzes git staged changes and generates meaningful commit messages
 */

const { execSync } = require("child_process");
const fs = require("fs");
const path = require("path");

const COMMIT_TYPES = {
  feat: "A new feature",
  fix: "A bug fix",
  docs: "Documentation only changes",
  style: "Changes that don't affect code meaning",
  refactor: "Code change that neither fixes a bug nor adds a feature",
  perf: "Code change that improves performance",
  test: "Adding or updating tests",
  chore: "Changes to build process or dependencies",
  ci: "CI configuration changes",
};

function getDiff() {
  try {
    return execSync("git diff --cached --name-only", { encoding: "utf-8" });
  } catch (e) {
    console.error("Error: Not a git repository or no staged changes");
    process.exit(1);
  }
}

function getDetailedDiff() {
  try {
    return execSync("git diff --cached", { encoding: "utf-8" });
  } catch (e) {
    return "";
  }
}

function analyzeChanges(files, diff) {
  const fileList = files.trim().split("\n").filter(Boolean);
  const stats = {
    added: 0,
    removed: 0,
    modified: 0,
    testFiles: 0,
    docFiles: 0,
    types: new Set(),
  };

  fileList.forEach((file) => {
    if (file.includes(".test.") || file.includes(".spec.")) stats.testFiles++;
    if (file.match(/\.(md|txt|rst)$/i)) stats.docFiles++;

    if (file.includes("package.json")) stats.types.add("chore");
    if (file.match(/\.(js|ts|jsx|tsx|py|java)$/i)) stats.types.add("feat");
  });

  const addedLines = (diff.match(/^\+/gm) || []).length - fileList.length;
  const removedLines = (diff.match(/^-/gm) || []).length - fileList.length;

  return {
    files: fileList,
    fileCount: fileList.length,
    stats: {
      addedLines,
      removedLines,
      testFiles: stats.testFiles,
      docFiles: stats.docFiles,
    },
    types: Array.from(stats.types),
  };
}

function generateMessage(analysis) {
  const { fileCount, stats, files, types } = analysis;

  // Determine commit type
  let commitType = "chore";
  if (stats.testFiles > 0) commitType = "test";
  else if (stats.docFiles === fileCount) commitType = "docs";
  else if (types.includes("feat")) commitType = "feat";

  // Generate subject based on changes
  let subject = "";
  const isSmallChange = fileCount <= 3;

  if (stats.docFiles === fileCount) {
    subject = `${commitType}: update documentation`;
  } else if (stats.testFiles > 0) {
    subject = `${commitType}: add/update tests`;
  } else if (stats.addedLines > stats.removedLines * 2) {
    subject = `${commitType}: add new functionality`;
  } else if (stats.removedLines > stats.addedLines) {
    subject = `${commitType}: remove unused code`;
  } else if (isSmallChange && fileCount === 1) {
    subject = `${commitType}: update ${path.basename(files[0])}`;
  } else {
    subject = `${commitType}: refactor code structure`;
  }

  // Generate body
  let body = `\nFiles changed: ${fileCount}\n`;
  body += `Lines added: ${stats.addedLines}\n`;
  body += `Lines removed: ${stats.removedLines}\n`;

  if (fileCount <= 5) {
    body += `\nModified files:\n`;
    files.forEach((file) => {
      body += `- ${file}\n`;
    });
  }

  return subject + body;
}

function main() {
  console.log("📝 Analyzing staged changes...\n");

  const files = getDiff();
  if (!files.trim()) {
    console.error("Error: No staged changes found. Stage your changes first with: git add .");
    process.exit(1);
  }

  const diff = getDetailedDiff();
  const analysis = analyzeChanges(files, diff);
  const message = generateMessage(analysis);

  console.log("═══════════════════════════════════════");
  console.log("✨ Generated Commit Message:");
  console.log("═══════════════════════════════════════\n");
  console.log(message);
  console.log("\n═══════════════════════════════════════");

  // Ask user if they want to commit
  const readline = require("readline");
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  rl.question(
    "\n✅ Do you want to commit with this message? (y/n): ",
    (answer) => {
      rl.close();
      if (answer.toLowerCase() === "y") {
        try {
          execSync("git commit -m " + JSON.stringify(message), {
            stdio: "inherit",
          });
          console.log("\n✓ Commit created successfully!");
        } catch (e) {
          console.error("Error creating commit");
          process.exit(1);
        }
      } else {
        console.log("\nCommit cancelled.");
      }
    }
  );
}

main();
